### PR TITLE
Make `MemoryView` Generic, make `cast` accurate

### PR DESCRIPTION
**Pull request by max-muoto**
As detailed in https://github.com/python/typeshed/issues/8182, the stubs for memoryview currently represent a sequence of ints, while at runtime this vary across ints, bytes, floats, and booleans at runtime after casting.

This PR fixes this by making `MemoryView` generic, while using 3.13 typevar defaults to avoid any type of large-scale regression, and adding appropiate overloads when casting.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Sounds like you found a mypy bug! Would be good to minimize it and report it over on the mypy repo.

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -823,8 +824,12 @@ class bytearray(MutableSequence[int]):
     def __buffer__(self, flags: int, /) -> memoryview: ...
     def __release_buffer__(self, buffer: memoryview, /) -> None: ...
 
+_IntegerFormats: TypeAlias = Literal[
```
Comment (max-muoto) [inline stdlib/builtins.pyi:827]: Let me know if we'd prefer these to be inline for the sake of IDEs, seemed harder to maintain so just decided to place them outside.

**Comment by max-muoto** [comment]
Most of the errors:

```
- aioredis/connection.py:933: error: Unsupported right operand type for in ("bytes | memoryview | int")  [operator]
+ aioredis/connection.py:933: error: Unsupported right operand type for in ("bytes | memoryview[int] | int")  [operator]
- aioredis/connection.py:934: error: Item "memoryview" of "bytes | memoryview | int" has no attribute "split"  [union-attr]
+ aioredis/connection.py:934: error: Item "memoryview[int]" of "bytes | memoryview[int] | int" has no attribute "split"  [union-attr]
- aioredis/connection.py:934: error: Item "int" of "bytes | memoryview | int" has no attribute "split"  [union-attr]
+ aioredis/connection.py:934: error: Item "int" of "bytes | memoryview[int] | int" has no attribute "split"  [union-attr]
- aioredis/client.py:4114: error: Incompatible types in assignment (expression has type "dict[bytes | str | memoryview, Any | None]", variable has type "dict[bytes | str | memoryview, Callable[[dict[str, str]], Awaitable[None]]]")  [assignment]
+ aioredis/client.py:4114: error: Incompatible types in assignment (expression has type "dict[bytes | str | memoryview[int], Any | None]", variable has type "dict[bytes | str | memoryview[int], Callable[[dict[str, str]], Awaitable[None]]]")  [assignment]
- aioredis/client.py:4158: error: Argument 1 to "update" of "MutableMapping" has incompatible type "dict[bytes | str | memoryview, Any | None]"; expected "SupportsKeysAndGetItem[bytes | str | memoryview, Callable[[dict[str, str]], Awaitable[None]]]"  [arg-type]
+ aioredis/client.py:4158: error: Argument 1 to "update" of "MutableMapping" has incompatible type "dict[bytes | str | memoryview[int], Any | None]"; expected "SupportsKeysAndGetItem[bytes | str | memoryview[int], Callable[[dict[str, str]], Awaitable[None]]]"  [arg-type]
- aioredis/client.py:4172: error: Incompatible types in assignment (expression has type "dict[bytes | str | memoryview, Callable[[dict[str, str]], Awaitable[None]]]", variable has type "dict[Any, Any | None]")  [assignment]
+ aioredis/client.py:4172: error: Incompatible types in assignment (expression has type "dict[bytes | str | memoryview[int], Callable[[dict[str, str]], Awaitable[None]]]", variable has type "dict[Any, Any | None]")  [assignment]
```

Are not net-new (just `memoryview` -> `memoryview[int]`).

For `web.py` in `aiohttp`, 

```python
class TCPSite(BaseSite):
    __slots__ = ("_host", "_port", "_reuse_address", "_reuse_port")

    def __init__(
        self,
        runner: "BaseRunner",
        host: Optional[str] = None,
        port: Optional[int] = None,
        *,
```

We see that the `isinstance` check should only narrow to `str` assuming `TCPSite` has accurate types.

**Comment by max-muoto** [comment]
> Sounds like you found a mypy bug! Would be good to minimize it and report it over on the mypy repo.

Are you referring to one of the previous crashes, or something else?

**Comment by JelleZijlstra** [comment]
> > Sounds like you found a mypy bug! Would be good to minimize it and report it over on the mypy repo.
> 
> Are you referring to one of the previous crashes, or something else? Or perhaps the `aiohttp` error not getting caught before.

I'm referring to the crash.

**Comment by max-muoto** [comment]
> > > Sounds like you found a mypy bug! Would be good to minimize it and report it over on the mypy repo.
> > 
> > 
> > Are you referring to one of the previous crashes, or something else? Or perhaps the `aiohttp` error not getting caught before.
> 
> I'm referring to the crash.

Sounds good, I'll put up an issue in that case.

**Comment by max-muoto** [comment]
Just wanted to bump this, and see if anyone was interested in reviewing?

**Pull request merged**